sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceEventClient.java

```diff
@@ -0,0 +1,402 @@
+            String frameworkName,
+            SchedulerConfig schedulerConfig,
+            MultiServiceManager multiServiceManager,
+            Collection<Object> customEndpoints,
+            UninstallCallback uninstallCallback) {
+        this.frameworkName = frameworkName;
+        this.schedulerConfig = schedulerConfig;
+        this.multiServiceManager = multiServiceManager;
+        this.customEndpoints = customEndpoints;
+        this.uninstallCallback = uninstallCallback;
+
+        if (schedulerConfig.isUninstallEnabled()) {
+            this.deregisterStep = new DeregisterStep();
+            this.uninstallPlan = Optional.of(
+                    new DefaultPlan(Constants.DEPLOY_PLAN_NAME, Collections.singletonList(
+                            new DefaultPhase(
+                                    "deregister-framework",
+                                    Collections.singletonList(deregisterStep),
+                                    new SerialStrategy<>(),
+                                    Collections.emptyList()))));
+        } else {
+            this.deregisterStep = null;
+            this.uninstallPlan = Optional.empty();
+        }
+    }
+
+    @Override
+    public void registered(boolean reRegistered) {
+        multiServiceManager.registered(reRegistered);
+    }
+
+    @Override
+    public void unregistered() {
+        if (!uninstallPlan.isPresent()) {
+            // This should have only happened after we returned OfferResponse.finished() below
+            throw new IllegalStateException("unregistered() called, but the we are not uninstalling");
+        }
+        deregisterStep.setComplete();
+    }
+
+    /**
+     * Forwards the provided offer(s) to all enclosed services, seeing which services are interested in them.
+     *
+     * TODO(data-agility): Lots of opportunities to optimize this. Needs benchmarks. For example:
+     * <ul>
+     * <li>- Hide reserved resources from services that they don't belong to</li>
+     * <li>- Forward the offers to a random ordering of services to avoid some services starving others</li>
+     * <li>- Distribute the offers across all services in parallel (optimistic offers)</li>
+     * <li>- ... Pretty much anything that you could see Mesos itself doing.</li>
+     * </ul>
+     */
+    @Override
+    public OfferResponse offers(Collection<Protos.Offer> offers) {
+        // If uninstalling, then finish uninstall. Otherwise decline short.
+        boolean noClients = false;
+        // Decline short.
+        boolean anyClientsNotReady = false;
+
+        List<OfferRecommendation> recommendations = new ArrayList<>();
+        List<Protos.Offer> remainingOffers = new ArrayList<>();
+        remainingOffers.addAll(offers);
+
+        Collection<String> finishedServices = new ArrayList<>();
+        Collection<String> uninstalledServices = new ArrayList<>();
+
+        Collection<AbstractScheduler> services = multiServiceManager.lockAndGetServices();
```

Comment:
Does holding this lock imply that add/remove/update operations on services are impossible now?  Do we want to lock, get copies, and then release the lock right away?  Do the add/remove/update operations just block?  It might be ok for now but this might be another place we want to file a JIRA to follow up and make sure we allow responsive API calls to clients.

Reply:
MultiServiceManager supports separate read (shared) vs read/write (exclusive) locking. Exclusive locks are only used when actually adding/removing a service from the set. All other access is only grabbing a shared lock.

This call is only getting a shared lock, so other shared-lock operations (like most API calls) can occur against any/all services at the same time as this is occurring.